File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.js
const getExpiredDate = (date, month) => {
    let d = new Date(date);
    d.setMonth(d.getMonth() + month);  
    d.setDate(d.getDate() - 1);        
    return d;
}

const isExpired = (target, today) => {
    return target < today;
}

function solution(today, terms, privacies) {
    const termsWithPeriod = {};
    const t = new Date(today);

    for (const term of terms){
        const [termName, period] = term.trim().split(' ');
        termsWithPeriod[termName] = parseInt(period);
    }

    const answer = [];

    privacies.forEach((privacy, idx) => {
        const [date, termName] = privacy.trim().split(' ');
        const period = termsWithPeriod[termName];
        const expiredDate = getExpiredDate(date, period);
        if (isExpired(expiredDate, t)) {
            answer.push(idx + 1);
        }
    });

    return answer;
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def get_day(date):
    yy,mm,dd = date.split('.')
    return int(dd) + int(mm)*28 + int(yy)*12*28

def solution(today, terms, privacies):
    today_days = get_day(today)
    
    terms_period_dict = {}
    for t in terms:
        term, expiration_period = t.split()
        terms_period_dict[term] = int(expiration_period) * 28
    
    answer = []
    for idx, p in enumerate(privacies):
        start, term_type = p.split()
        start_days = get_day(start)
        end_days = start_days + terms_period_dict[term_type] - 1
        if today_days > end_days:
            answer.append(idx + 1)
    
    return answer